impeller/renderer/backend/metal/vertex_descriptor_mtl.mm

```diff
@@ -188,29 +188,52 @@ static MTLVertexFormat ReadStageInputFormat(const ShaderStageIOSlot& input) {
   return true;
 }
 
+void VertexDescriptorMTL::SetInterleavedVertexData(bool value) {
+  interleaved_vertex_data_ = value;
+}
+
 MTLVertexDescriptor* VertexDescriptorMTL::GetMTLVertexDescriptor() const {
   auto descriptor = [MTLVertexDescriptor vertexDescriptor];
 
-  const size_t vertex_buffer_index =
-      VertexDescriptor::kReservedVertexBufferIndex;
-
-  size_t offset = 0u;
-  for (const auto& input : stage_inputs_) {
-    auto attrib = descriptor.attributes[input.location];
-    attrib.format = input.format;
-    attrib.offset = offset;
-    // All vertex inputs are interleaved and tightly packed in one buffer at a
-    // reserved index.
-    attrib.bufferIndex = vertex_buffer_index;
-    offset += input.length;
-  }
-
-  // Since it's all in one buffer, indicate its layout.
-  auto vertex_layout = descriptor.layouts[vertex_buffer_index];
-  vertex_layout.stride = offset;
-  vertex_layout.stepRate = 1u;
-  vertex_layout.stepFunction = MTLVertexStepFunctionPerVertex;
+  if (interleaved_vertex_data_) {
+    const size_t vertex_buffer_index =
+        VertexDescriptor::kReservedVertexBufferIndex;
+
+    size_t offset = 0u;
+    for (const auto& input : stage_inputs_) {
+      auto attrib = descriptor.attributes[input.location];
+      attrib.format = input.format;
+      attrib.offset = offset;
+      // All vertex inputs are interleaved and tightly packed in one buffer at a
+      // reserved index.
+      attrib.bufferIndex = vertex_buffer_index;
+      offset += input.length;
+    }
 
+    // Since it's all in one buffer, indicate its layout.
+    auto vertex_layout = descriptor.layouts[vertex_buffer_index];
+    vertex_layout.stride = offset;
+    vertex_layout.stepRate = 1u;
+    vertex_layout.stepFunction = MTLVertexStepFunctionPerVertex;
+
+  } else {
+    // each piece of vertex data is in its own buffer. They are assigned
+    // starting at kReservedVertexBufferIndex counting backwards.
+    size_t vertex_buffer_index = VertexDescriptor::kReservedVertexBufferIndex;
+
+    for (const auto& input : stage_inputs_) {
+      auto attrib = descriptor.attributes[input.location];
+      attrib.format = input.format;
+      attrib.offset = 0u;
+      attrib.bufferIndex = vertex_buffer_index;
+
+      auto vertex_layout = descriptor.layouts[vertex_buffer_index];
+      vertex_layout.stride = input.length;
+      vertex_layout.stepRate = 1u;
+      vertex_layout.stepFunction = MTLVertexStepFunctionPerVertex;
```

Comment:
Can we research a design where the vertex layout is always provided to Impeller's HAL rather than having any implicit/hardcoded layouts enforced by the backends?
I think we'll inevitably need fine control over layout in some situations and I'm a bit concerned that adding a second hardcoded layout style will increase the difficulty of backing out of this.

For example, something like:
1. The pipeline stores an attribute list.
2. ImpellerC generates a default attribute list in the vertex stage structs with the staggered layout.
3. This default layout is copied to the pipeline unless overridden in the ctor.
4. When using a custom layout, the user is entirely responsible for packing the set of vertex buffers correctly and binding them on the raster Command. When not using a custom layout, they can safely use the generated header utilities and everything works fine.